src/sorting/tim_sort.rs

```diff
@@ -94,42 +124,80 @@ pub fn tim_sort(arr: &mut Vec<i32>, n: usize) {
 #[cfg(test)]
 mod tests {
     use super::*;
-    use crate::sorting::have_same_elements;
-    use crate::sorting::is_sorted;
+    use crate::sorting::{have_same_elements, is_sorted};
+
+    #[test]
+    fn min_run_length_returns_correct_value() {
+        assert_eq!(min_run_length(0), 0);
+        assert_eq!(min_run_length(10), 10);
+        assert_eq!(min_run_length(33), 17);
+        assert_eq!(min_run_length(64), 16);
+    }
+
+    #[test]
+    fn insertion_sort_sorts_array_correctly() {
+        let mut array = vec![3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5];
+        insertion_sort(&mut array);
+        assert_eq!(array, vec![1, 1, 2, 3, 3, 4, 5, 5, 5, 6, 9]);
+    }
+
+    #[test]
+    fn merge_left_and_right_subarrays_into_array() {
+        let mut array = vec![0, 2, 4, 1, 3, 5];
+        merge(&mut array, 0, 2, 5);
+        assert_eq!(array, vec![0, 1, 2, 3, 4, 5]);
+    }
+
+    #[test]
+    fn merge_with_empty_left_subarray() {
+        let mut array = vec![1, 2, 3];
+        merge(&mut array, 0, 0, 2);
+        assert_eq!(array, vec![1, 2, 3]);
+    }
+
+    #[test]
+    fn merge_with_empty_right_subarray() {
+        let mut array = vec![1, 2, 3];
+        merge(&mut array, 0, 2, 2);
+        assert_eq!(array, vec![1, 2, 3]);
+    }
+
+    #[test]
+    fn merge_with_empty_left_and_right_subarrays() {
+        let mut array = vec![1, 2, 3];
+        merge(&mut array, 1, 0, 0);
+        assert_eq!(array, vec![1, 2, 3]);
+    }
 
     #[test]
-    fn basic() {
+    fn tim_sort_sorts_basic_array_correctly() {
         let mut array = vec![-2, 7, 15, -14, 0, 15, 0, 7, -7, -4, -13, 5, 8, -14, 12];
         let cloned = array.clone();
-        let arr_len = array.len();
-        tim_sort(&mut array, arr_len);
+        tim_sort(&mut array);
         assert!(is_sorted(&array) && have_same_elements(&array, &cloned));
     }
 
     #[test]
-    fn empty() {
+    fn tim_sort_handles_empty_array() {
         let mut array = Vec::<i32>::new();
         let cloned = array.clone();
-        let arr_len = array.len();
-        tim_sort(&mut array, arr_len);
+        tim_sort(&mut array);
         assert!(is_sorted(&array) && have_same_elements(&array, &cloned));
     }
 
     #[test]
-    fn one_element() {
+    fn tim_sort_handles_single_element_array() {
         let mut array = vec![3];
         let cloned = array.clone();
-        let arr_len = array.len();
-        tim_sort(&mut array, arr_len);
+        tim_sort(&mut array);
         assert!(is_sorted(&array) && have_same_elements(&array, &cloned));
     }
 
     #[test]
-    fn pre_sorted() {
+    fn tim_sort_handles_pre_sorted_array() {
         let mut array = vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9];
         let cloned = array.clone();
-        let arr_len = array.len();
-        tim_sort(&mut array, arr_len);
+        tim_sort(&mut array);
         assert!(is_sorted(&array) && have_same_elements(&array, &cloned));
     }
```

Comment:
I would suggest to _parametrize_ that by using:

```rust
    macro_rules! test_tim_sort {
        ($($name:ident: $input:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let mut array = $input;
                    let cloned = array.clone();
                    tim_sort(&mut array);
                    assert!(is_sorted(&array) && have_same_elements(&array, &cloned));
                }
            )*
        }
    }

    test_tim_sort! {
        sorts_basic_array_correctly: vec![-2, 7, 15, -14, 0, 15, 0, 7, -7, -4, -13, 5, 8, -14, 12],
        sorts_long_array_correctly: vec![-2, 7, 15, -14, 0, 15, 0, 7, -7, -4, -13, 5, 8, -14, 12, 5, 3, 9, 22, 1, 1, 2, 3, 9, 6, 5, 4, 5, 6, 7, 8, 9, 1],
        handles_empty_array: Vec::<i32>::new(),
        handles_single_element_array: vec![3],
        handles_pre_sorted_array: vec![0, 1, 2, 3, 4, 5, 6, 7, 8, 9],
    }
```

Please notice an additional test case.